### PR TITLE
Invoke J9AllocateObject without acquiring liveVirtualThreadListMutex

### PR DESCRIPTION
**Pull request by babsingh**
This change fixes the below deadlock.

Thread 1 wants exclusive VM access but it is unable to acquire
exclusive VM access because Thread 2 also holds VM access.
Thread 2 is waiting to acquire liveVirtualThreadListMutex and
unable to proceed because Thread 1 holds liveVirtualThreadListMutex.
This prevents both threads to proceed and leads to a deadlock.

```
Thread 1: JVM_VirtualThreadMountEnd
           -> Acquires VM Access and liveVirtualThreadListMutex
           -> J9AllocateObject
           -> Multiple GC frames
           -> acquireExclusiveVMAccess

Thread 2: JVM_VirtualThreadMountEnd
           -> Acquires VM Access
           -> Blocked to enter liveVirtualThreadListMutex
```

Related: https://github.com/eclipse-openj9/openj9/issues/16340#issuecomment-1324095402

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
jenkins test sanity amac jdk19

**Comment by babsingh** [comment]
@gacholio Can you please review this PR?

I tried the startup solution in https://github.com/eclipse-openj9/openj9/pull/16384. It seems too early for some dependencies. This prevents us to use the startup solution.

fyi @fengxue-IS @tajila

File: runtime/j9vm/javanextvmi.cpp [hunk 1]
```diff
@@ -305,12 +305,31 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
...
+	j9object_t threadObj = NULL;
 
 	Trc_SC_VirtualThreadMountEnd_Entry(currentThread, thread, firstMount);
 
 	vmFuncs->internalEnterVMFromJNI(currentThread);
+	/* On some occassions, J9AllocateObject acquires exclusive VM access. It is invoked without acquiring
+	 * liveVirtualThreadListMutex to prevent a deadlock.
+	 */
+	if (firstMount && (NULL == vm->liveVirtualThreadList)) {
+		J9Class *virtualThreadClass = J9OBJECT_CLAZZ(currentThread, J9_JNI_UNWRAP_REFERENCE(thread));
```
[the remark this answers is not in the document]
Comment (tajila) [reply runtime/j9vm/javanextvmi.cpp:318]: actually ignore this, its in https://github.com/eclipse-openj9/openj9/pull/16365

[the remark this answers is not in the document]
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp:318]: a similar assertion is being added in https://github.com/eclipse-openj9/openj9/pull/16365. i will wait for it to be merged and rebase.

File: runtime/j9vm/javanextvmi.cpp [hunk 2]
```diff
@@ -305,12 +305,31 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
...
+	j9object_t rootVirtualThread = NULL;
+	j9object_t threadObj = NULL;
 
 	Trc_SC_VirtualThreadMountEnd_Entry(currentThread, thread, firstMount);
 
 	vmFuncs->internalEnterVMFromJNI(currentThread);
+	/* On some occassions, J9AllocateObject acquires exclusive VM access. It is invoked without acquiring
+	 * liveVirtualThreadListMutex to prevent a deadlock.
+	 */
+	if (firstMount && (NULL == vm->liveVirtualThreadList)) {
```
[the remark this answers is not in the document]
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp]: updated.

File: runtime/j9vm/javanextvmi.cpp [hunk 3]
```diff
@@ -326,34 +327,41 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
...
+		Assert_SC_true(firstMount);
+		J9Class *virtualThreadClass = J9OBJECT_CLAZZ(currentThread, J9_JNI_UNWRAP_REFERENCE(thread));
+		J9MemoryManagerFunctions *mmFuncs = vm->memoryManagerFunctions;
+
+		/* Allocate empty virtual thread and create a global reference to it as root for the linked list.
+		 * This prevents the root reference from becoming stale if the GC moves the object.
+		 */
+		rootVirtualThread = mmFuncs->J9AllocateObject(currentThread, virtualThreadClass, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+		if (NULL == rootVirtualThread) {
+			vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
```
Comment (gacholio) [inline runtime/j9vm/javanextvmi.cpp:343]: This is a heap OOM, not native.

**Comment by babsingh** [comment]
Rebased since https://github.com/eclipse-openj9/openj9/pull/16365 was merged.

File: runtime/j9vm/javanextvmi.cpp (hunk 3 above)
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp]: Updated both locations.

Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp]: Will the other `j9jni_createGlobalRef` location also be a heap OOM? If not, I will revert that change.

Comment (gacholio) [reply runtime/j9vm/javanextvmi.cpp]: Yes - refs are allocated in native memory.

File: runtime/j9vm/javanextvmi.cpp [hunk 4]
```diff
@@ -326,34 +327,41 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
...
-
-					/* Set linkNext/linkPrevious to itself. */
-					J9OBJECT_OBJECT_STORE(currentThread, rootVirtualThread, vm->virtualThreadLinkNextOffset, rootVirtualThread);
-					J9OBJECT_OBJECT_STORE(currentThread, rootVirtualThread, vm->virtualThreadLinkPreviousOffset, rootVirtualThread);
-				} else {
-					vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
-				}
-			} else {
+			jobject globalRef = vmFuncs->j9jni_createGlobalRef((JNIEnv *)currentThread, rootVirtualThread, JNI_FALSE);
+			if (NULL == globalRef) {
```
[the remark this answers is not in the document]
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp:356]: Updated.

File: runtime/j9vm/javanextvmi.cpp [hunk 5]
```diff
@@ -326,34 +327,41 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
...
-				}
-			} else {
-				vmFuncs->setHeapOutOfMemoryError(currentThread);
+			jobject globalRef = vmFuncs->j9jni_createGlobalRef((JNIEnv *)currentThread, rootVirtualThread, JNI_FALSE);
+			if (NULL == globalRef) {
+				vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+				goto release2;
 			}
+
+			vm->liveVirtualThreadList = (j9object_t *)globalRef;
```
Comment (gacholio) [inline runtime/j9vm/javanextvmi.cpp:361]: Just use `rootVirtualThread` here.

Comment (gacholio) [reply runtime/j9vm/javanextvmi.cpp:361]: Ignore - misread again.

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk19

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk19

**Comment by babsingh** [comment]
Re-added the tracepoint which was introduced in 

The old PR build should still be good: https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/3090/.

File: runtime/j9vm/javanextvmi.cpp [hunk 6]
```diff
@@ -326,34 +327,42 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
...
+
+		/* Allocate empty virtual thread and create a global reference to it as root for the linked list.
+		 * This prevents the root reference from becoming stale if the GC moves the object.
+		 */
+		rootVirtualThread = mmFuncs->J9AllocateObject(currentThread, virtualThreadClass, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+		if (NULL == rootVirtualThread) {
+			vmFuncs->setHeapOutOfMemoryError(currentThread);
+			goto release1;
+		}
+		J9VMJAVALANGVIRTUALTHREAD_SET_STATE(currentThread, rootVirtualThread, J9VM_VIRTUALTHREAD_ROOT_NODE_STATE);
```
Comment (babsingh) [inline runtime/j9vm/javanextvmi.cpp:346]: @keithc-ca https://github.com/eclipse-openj9/openj9/pull/16365#discussion_r1037263435 is addressed here.

**Comment by tajila** [comment]
jenkins test sanity amac jdk19

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins test sanity amac jdk19

**Comment by tajila** [comment]
s390x build failed with infra issue

```
11:19:32  Starting building: [Test_openjdk19_j9_sanity.functional_s390x_linux_Personal #50](https://openj9-jenkins.osuosl.org/job/Test_openjdk19_j9_sanity.functional_s390x_linux_Personal/50/)
Creating placeholder flownodes because failed loading originals.
java.io.IOException: Tried to load head FlowNodes for execution Owner[Pipeline_Build_Test_JDK19_s390x_linux/257:Pipeline_Build_Test_JDK19_s390x_linux #257] but FlowNode was not found in storage for head id:FlowNodeId 1:110
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.initializeStorage(CpsFlowExecution.java:679)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:716)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:701)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:560)
	at hudson.model.RunMap.retrieve(RunMap.java:233)
	at hudson.model.RunMap.retrieve(RunMap.java:61)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:650)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:632)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:530)
	at hudson.model.RunMap.getById(RunMap.java:213)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.run(WorkflowRun.java:949)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.get(WorkflowRun.java:961)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$1.computeNext(FlowExecutionList.java:76)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$1.computeNext(FlowExecutionList.java:68)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:146)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:141)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$ItemListenerImpl.onLoaded(FlowExecutionList.java:197)
	at jenkins.model.Jenkins.<init>(Jenkins.java:1036)
	at hudson.model.Hudson.<init>(Hudson.java:86)
	at hudson.model.Hudson.<init>(Hudson.java:82)
	at hudson.WebAppMain$3.run(WebAppMain.java:247)
Finished: FAILURE
```

**Comment by tajila** [comment]
jenkins test sanity zlinux jdk19

**Pull request merged**